### ISSUE TITLE
Listing ports in docker-compose not in ascending order causes container reboot loop

### ISSUE THREAD

**Issue by yutter** [balena-io/balena-supervisor#824]
If I do not list the ports in the docker-compose file in ascending order the container starts / dies / starts in a continuous loop. This issue started when I upgraded my BalenaOS from 2.13.x to 2.26.x
The issue is the same on my RaspberryPI 3s and UP2s

This works in all BalenaOS versions:
    ports:
      - 80:80
      - 443:443
      - 502:502
      - 4840:4840
      - 8883:8883
      - 10001:10001
      - 30815:30815
      - 30883:30883

This causes the reboot loop in v2.26.x but worked fine in v2.13.x:
    ports:
      - 80:80
      - 443:443
      - 502:502
      - 4840:4840
      - 8883:8883
      - 30815:30815
      - 30883:30883
      - 10001:10001

I have also noticed that when the container gets into this reboot loop it does seem to get out of it and run properly when I am downloading a new container image! No idea why that would happen

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_g9c3)

**Comment by yutter** [comment]
Also just noticed that the BalenaCloud dashboard shows the correct commit IDs for the containers but shows 'factory build' for the device when it is in this loop. After rebuilding my project with ports in ascending order the device commit ID is updated correctly.

FYI I am running two containers in my project and the other container which did not expose any ports ran fine

**Comment by CameronDiver** [comment]
I've managed to reproduce an issue which may be the same or not, I'm not sure. Interestingly this is only exhibited when the code is sent via the builder, and not with the new beta local mode. This implies a version of the composition parser might be at fault, but I'll continue to look into this. I have a load of other stuff to do today, but this is next on my priority list.

> Also just noticed that the BalenaCloud dashboard shows the correct commit IDs for the containers but shows 'factory build' for the device when it is in this loop. After rebuilding my project with ports in ascending order the device commit ID is updated correctly.

This is expected, the supervisor will only report having moved to another release once it has correctly "settled", running the release as it was intended.

**Comment by agherzan** [comment]
I managed to reproduce this on 2.27.0 RaspberryPi 3 after updating the supervisor from 7.16.6 to 8.3.5 with hostOS updates.

**Comment by janvda** [comment]
I have also faced this issue and reordering the ports fixed this issue.

Note that I have lost many hours troubleshooting this issue especially as the logs don't give any indication that there is an error with the order in which the port numbers are specified.
See : https://forums.balena.io/t/container-in-endless-installing-starting-installing-stopping-loop-since-upgrade-from-resin-os-2-15-1-rev2-to-balenaos-2-27-0-rev1/4790

So can the balena team for the benefit of others give some priority to this issue (or at least write an appropriate error in the log files).

**Comment by CameronDiver** [comment]
Apologies for the delay in getting to this due to the holidays - I've looked into this and it comes from the fact that docker will always report back the ports in ascending order, so when the supervisor compares the running container with the one requested in the release, it's always different. This is top of my priority list and should be fixed in the next day or so.

**Comment by kkom** [comment]
Thanks @CameronDiver for the update!

I've also encountered the issue, but wasn't able to apply the workaround. It may have been due to the fact that I published UDP ports as well (I am not sure what's the right way to sort them).

Anyway, I'm just letting you know to be mindful of that scenario as well.

When could your fix be tested in production?

**Comment by kkom** [comment]
Cool! Thanks for the info!

Can we also take a chance to rethink how this error was surfaced to the user?

I'm not suggesting to print a detailed message about misaligned ports -- that's definitely too detailed.

I'm suggesting to surface the higher-level fact that the restart loop comes from the supervisor, as opposed to from some bug inside the container.

Right now, the error message was very unfriendly, see these two quotes:

> Note that I have lost many hours troubleshooting this issue especially as the logs don't give any indication that there is an error with the order in which the port numbers are specified.

( https://github.com/balena-io/balena-supervisor/issues/824#issuecomment-450763280 )

> It is stuck in an endless Installing - Starting - Installing - Stopping loop of the service `influxdb sha256:739ccc7e7ca4d47646209f454c24254a641ad438ff74cc5f2c2ff680421dbfbc`
. I have no clue why.

( https://forums.balena.io/t/container-in-endless-installing-starting-installing-stopping-loop-since-upgrade-from-resin-os-2-15-1-rev2-to-balenaos-2-27-0-rev1/4790 )

What do you think?

**Comment by CameronDiver** [comment]
@kkom There is actually a load of information printed to the journal by the supervisor in these situations, available with `journalctl -f -u resin-supervisor` from the host. This information should eventually makes it's way to the dashboard (albeit in a much more formatted form) but currently it's only available in the journal. In situations like this, the logs are invaluable, as they tend to tell you exactly why the supervisor has deemed that it needs to replace a running container.

I'll make an issue to track the implemntation of a system which allows these data to be displayed in the dashboard too, as I do agree the UX in these cases is not great.

**Comment by kazazes** [comment]
This issue is confirmed to still be present as of 2.29.2. It took a few hours to find this issue, but ordering the ports solved the problem. This was one of ~10 containers.

```
Mar 24 13:59:41 495bb98 balenad[655]: [2019-03-24T13:59:41.725Z] Replacing container for service datarouter because of config changes:
Mar 24 13:59:41 495bb98 resin-supervisor[20627]: [2019-03-24T13:59:41.725Z] Replacing container for service datarouter because of config changes:
```

```yaml
datarouter:
    container_name: datarouter
    build:
      context: ./host-images/data-router
    volumes:
      - "resin-data:/data"
    ports:
      - "1883"
      - "4000"
    environment:
      - NODE_ENV=production
      - GRAPHQL_PORT=4000
      - MQTT_PORT=1883
      - GRAPHQL_WS=wss://edge.sibyl.vision/graphql/subscriptions
      - GRAPHQL_HTTP=https://edge.sibyl.vision/graphql/
    depends_on:
      - redis
```

**Comment by kkom** [comment]
Thank you @kazazes for re-reporting! I would love to get to the bottom of this to be able to robustly host a UniFi controller on a Raspi -- it needs to open a ton of ports!

**Comment by CameronDiver** [comment]
@kazazes @kkom this was fixed in balena-supervisor version v9.7.2 (https://github.com/balena-io/balena-supervisor/blob/master/CHANGELOG.md#972---2019-02-09) which wasn't available until balenaOS 2.30.x.

Could you retry with this version and let me know if you're still experiencing errors? FWIW I tested the syntax of your service myself with the latest supervisor+OS and it worked fine.
